### DISCUSSION TITLE
Functional/selector atoms

### DISCUSSION THREAD

**Discussion by norswap** [pmndrs/jotai#2300] [Ideas] [answered]
I'd like to have an atom that defines a function (i.e. where `get` can take parameters): like so:

```typescript
const array = atom([])
const indexArray = atom((get, index) => get(array)[index])
```

And could be used as `const value = useAtom(indexArray, index)`

It seems like this would be useful every you're time you're dealing with a data structure containing multiple members (array, dict), where a component only needs access to one of the members.

Right now, you're forced to use the whole data structure as atom, so you'll rerender every time the data changes, not only the member you're interested in.

There is a somewhat similar question [here](https://www.reddit.com/r/reactjs/comments/nfm3qp/help_with_jotai_making_selector/) which proposes to create the derived atom lazily.

So in this case:

```
function indexArray(index) {
   return atom(get) = > get(array)[index]
}
```

and then `useAtom(useMemo(() => indexArray(index), [index])`

This works, but I'm worried about cluttering the app state with atoms that later become irrelevant. As far as I understand, the `array` atom will keep a reference to each `indexArray` atom created this way, even after they are no longer needed.

1. Is there a way to achieve this today while sidestepping the issue with garbage atoms accumulating?
2. Could the feature be considered for inclusion?

**Comment by norswap** [comment]
Okay, I had a look at the code (should have done that before), and it seems like atoms are only kept around when mounted in React (`useAtom` causes a call to `store.sub` which causes the atom to be registered, while the returned `unsub` deregisters it). So creating these atoms lazily should work.

**Comment by dai-shi** [comment, marked as answer]
We have similar discussion before and it's technically out of the scope of Jotai core.
useMemo is a recommended solution in your safe. It's safe. Jotai store is only WeakMap based, and this allows atoms-in-atom pattern too. Creating a new atom is very lightweight. subscribe/unsubscribe comes with a cost, so avoid it unnecessarily.

We have a new library https://github.com/jotaijs/jotai-scope which is related. (but, not specific to your use case.)

Let me know if you have any other questions.
